core/third_party_libs/pue/pue-client.php

```diff
@@ -1032,6 +1032,25 @@ private function requestInfo(array $queryArgs = array())
                 if (isset($response->extra_notices)) {
                     $this->add_persistent_notice($response->extra_notices);
                 }
+                // This instance is for EE Core, we have a response from PUE so lets check if it contains PUE Plugin data.
+                if ($this->slug == 'event-espresso-core-reg' && isset($response->extra_data) && !empty($response->extra_data->plugins) ) {
+                    // Pull PUE pugin data from 'extra_data'.
+                    $plugins_array = json_decode($result['body'], true);
+                    $plugins = $plugins_array['extra_data']['plugins'];
+                    // Pull all of the add-ons EE has active and update the local latestVersion value of each of them.
+                    foreach (EE_Registry::instance()->addons as $addon) {
+                        if( isset($plugins[ $addon->getPueSlug() ]) ) {
+                            $addon_state = get_option('external_updates-' . $addon->getPueSlug());
+                            // If we don't have an addon state, get out we'll update it next time.
+                            if( empty($addon_state)) {
+                                continue;
+                            }
+                            // Have an addon state? Set the latestVersion value!
+                            $addon_state->latestVersion = $plugins[$addon->getPueSlug()];
+                            update_option('external_updates-' . $addon->getPueSlug(), $addon_state);
```

Comment:
how does this ever get set the very first time?

initially this option will not exist in the db right? or was this something that already existed in PUE?

Reply:
It already exists within PUE, it's the option for each plugin that stores the update checks book-keeping ('checkPerion', 'update' details like the PUE download url for add-on etc), so it will already be set on any add-ons installed on the site, if it's not then on the next call home it will be.

Reply:
> It already exists within PUE

so this same property gets set via some other code in PUE?


> it's the option for each plugin that stores the update checks book-keeping

...mmm... sry... not understanding this sentence... what are the "update checks book-keeping" ???

> so it will already be set on any add-ons installed on the site

... ok... but when/where does that happen? Again, is there some other code that saves this option? If so then won't this conflict with that other logic? ie: two blocks of code updating an option that's intended to store the current state of affairs sounds potentially fragile.

> if it's not then on the next call home it will be.

But isn't THIS the code that does the calling home? if the option isn't set, then on the next call home, won't it return as soon as it hits this block:

```php
if( empty($addon_state)) {
	continue;
}
```

sry not trying to be a PITA, just trying to ensure that everything is as solid as it can be

Reply:
>so this same property gets set via some other code in PUE?

By property do you mean `$state->latestVersion` or `$state` as a whole?

`$state->latestVersion` is only set within this code and I chose this location because it is after PUE has set up `$state` for the current plugin, note that the code in question here only runs on EE core requests.

The only place that creates the 'state' object is in [checkForUpdates()](https://github.com/eventespresso/event-espresso-core/blob/ENH/reduce-pue-load/core/third_party_libs/pue/pue-client.php#L1662-L1686), which calls [requestUpdate(), which calls requestInfo()](https://github.com/eventespresso/event-espresso-core/blob/ENH/reduce-pue-load/core/third_party_libs/pue/pue-client.php#L1336-L1343) (which is where this code is)

That code to set `latestVersion` on each add-on `$state` can only run if we have a `$state` for that add-on, right? The call home happens if `$state->latestVerion` is empty or it is greater than the currently installed version. So I'm confirming we have a state object for the add-ons before setting `latestVersion` on them. The add-ons need to call home at least once and will do so when installed, that sets up `$state`, and it's set on the current request even before this code runs in checkForUpdates().

So, when we activate an add-on (let's use the calendar as an example), it runs through PUE and gets to `checkForUpdates()`, at that point it doesn't have a `$state` set yet.

Because it doesn't have a `$state`, my new version number check is skipped. `$state` is created and saved to the database (at that point is holds the current time in (`$state->lastCheck) and the currently installed version (`$state->checkedVersion`) and has a `update` property set to null (that property holds the update values WP needs to run an update on the plugin and is used only when an update is actually available).

The above is also what I meant about the 'book-keeping' regarding updates, `$state` stores when the last check was done, the update values etc, everything concerning this current plugins 'PUE State' is stored in this option.

> .. ok... but when/where does that happen? Again, is there some other code that saves this option? If so then won't this conflict with that other logic? ie: two blocks of code updating an option that's intended to store the current state of affairs sounds potentially fragile.

This is the only location that saves the `latestVersion` property on `$state` for each add-on that has a state.

I'm not sure I follow how this is fragile, should I not be updating an option after its been created?

Nowhere else uses `latestVersion`, they do use state but don't loop over the properties of it unsetting them etc.

> But isn't THIS the code that does the calling home? if the option isn't set, then on the next call home, won't it exit the loop as soon as it hits this block

The code to save `latestVersion` on each add-on only runs on a call home from core, so no.

Using the EE calendar add-on example from above, when you install that add-on it will call home and when it does it creates its `$state` and saves it, this code in question doesn't not run on that call as its not EE core. So then when core calls home it updates all add-ons states with their latest versions, if we don't have a state for EE Calendar something is up and I don't want to create one as PUE should be doing it.

The idea behind this method is to leave PUE to do whatever it does already unless `latestVersion` has been set and its value is equal to the current version (or less than but that's unlikely without people hacking stuff up so that's on them). `latestVersion` is only set on an add-on when we have a state for it (PUE did its thing as normal for that add-on) and EE4 core called home to pull the latestVersions and set them on the add-ons states.

> sry if I'm totally missing something, not trying to be a PITA, just trying to ensure that everything is as solid as it can be

I have 0 issues with you asking for more details on anything I've written, I learn a huge amount from you guys questioning what and why :)

Reply:
> By property do you mean $state->latestVersion or $state as a whole?

oh dang... no... sry that was a bad word choice... 
I meant **option** not property, as in the option that is saved using the key `'external_updates-' . $addon->getPueSlug()` (which I'll simply refer to as the "external_updates" option from here on in)

ok I shovelled through the PUE code some more after reading the above and think I grok what's going on...

So when each individual addon is registered, `EE_Register_Addon::load_pue_update()` instantiates an instance of `PluginUpdateEngineChecker` which at some point calls `checkForUpdates()`. That method will save the "external_updates" option for the add-on. Eventually the code gets to the new logic that you added in `requestInfo()`, and since it's not EE core, it will exit and NOT phone home. When EE core eventually gets around to `requestInfo()` it won't exit, and instead will loop through all of the registered addons, pull their "external_updates" option, which _should_ have already been set, and if needed, will phone home for updates.

Initially, when I saw `checkForUpdates()` doing this:

```php
update_site_option($this->optionName, $state);
```

and noticed that:

```php
$this->optionName = 'external_updates-' . $this->slug;
```

I was erringly thinking that the above only got set for EE core and NOT for addons because you had to add the `EE_Addon::getPueSlug()` method in order to get the slug for addons... so `'external_updates-' . $this->slug` couldn't be for addons because otherwise you would have used that for getting the addon slug. So that **HAD** to be just for EE core... see my confusion? Of course, `$this->slug` is ultimately the same thing as `EE_Addon::getPueSlug()` when the instance of `PluginUpdateEngineChecker` is for the same addon. So I was having a blonde moment :confused: 

So bottom line, I erringly thought that the only place "external_updates" option was getting saved for addons, was in your new code, but you **ONLY** did that if the option already existed, which is why I was wondering how it got created in the first place. Your statement that

> if it's not (set) then on the next call home it will be.

kinda reaffirmed my mistaken assumption the "external_updates" option was ONLY saved from within the new "call home" logic, whereas it is **actually** being saved in `checkForUpdates()` when the addon is registered... NOT when EE core phones home. I don't think you meant to imply how I interpreted things but it's all good now and i think I fully grok what's going on.

> I'm not sure I follow how this is fragile, should I not be updating an option after its been created?

it's not fragile to update an option after it has been created, else you could never update an option. My concern is because that option is being updated in two different places using potentially different information. If the information wasn't different, then there would be no reason to update the "external_updates" option for an addon from EE core, as the info that had already been saved by the addon would be the same. 

So this is my concern:

- on every request, `EE_Register_Addon::register()` is called for each addon, which in turn calls `_parse_pue_options()`
- the 'AHEE__EE_System__brew_espresso__after_pue_init' hook is set which is triggered in `EE_System::brew_espresso()`
- `EE_Register_Addon::load_pue_update()` runs for addon XYZ during its registration process
- `PluginUpdateEngineChecker::checkForUpdates()` is called 
- the `installed_version` is compared against the `latestVersion` and if we are already up to date, then we do nothing
- otherwise the "external_updates" option is saved with what we'll call "version A" for that info
- after all addons finish registering, EE core runs `PluginUpdateEngineChecker` for itself
- `PluginUpdateEngineChecker::requestInfo()` runs, calls home, and updates the `latestVersion` value for addon XYZ (cuz a new version is out)
- "external_updates" option is saved with what we'll call "version B" for that info

now here's where I'm imagining a potential issue...

- on the next request... if the user has not updated addon XYZ, the above process happens again...  which means that addon XYZ is going to overwrite the "version B" of the "external_updates" option because:

```php
            if( $this->slug !== 'event-espresso-core-reg' && !empty($state->latestVersion) ) {
                if( version_compare($this->_installed_version, $state->latestVersion, '>=') ) {
                    return;
                }
            }
```
(side note: the above conditionals can be combined into one big IF statement)

now because addon XYZ was not updated, the installed version is going to be LESS than the `latestVersion`, meaning we do NOT return from the above... and will continue on to resave the "external_updates" option, and essentially overwrite version B with version A.

So ultimately it seems like that the "external_updates" option is potentially going to get repeatedly saved and overwritten back n forth with different data , assuming the data is different during those two updates which I'm assuming is the case considering the amount of code that's in PUE, and also assuming that there isn't something else that I'm missing that happens somewhere else in the PUE code to prevent this.

Anyways... big ramble to explain what my concern was, but if the above is completely unwarranted worrying, then just disregard all of that : \

Reply:
Thanks for digging into it, I knida breezed over a little too much above forgetting that not everyone has been going over PUE recently :)

> now because addon XYZ was not updated, the installed version is going to be LESS than the latestVersion, meaning we do NOT return from the above... and will continue on to resave the "external_updates" option, and essentially overwrite version B with version A.

Ah, ok. I follow.

So my understanding of this is that at that point "external_updates" (`$state` in code) is going to include:

```
$state->lastCheck
$state->checkedVersion
$state->latestVersion
$state->update
```

Because there is an update available, `update` is going to be populated with various data but the actual contents are unimportant for this so just noting that its there.

`$state->latestVersion` was _only_ set in in the previous request with from core.

`$state->update` is set by the call to `requestUpdate()` in the `checkForUpdates()` method, however, right before that the option is saved.

Which means `checkForUpdates()` is going to pull the option (which again, in this request includes `lastestVersion`) update `lastCheck`, update `checkedVersion` and save the option back again (it does this before `requestUpdate()` and has a comment stating it does this just in case something goes wrong with the `requestUpdate()` call).

It then pulls an update from ee.com and saves/updates `$state->update` with whatever is returned here, then again, saves the option.

So I don't think that `latestVersion` (which is the important part of this new code, right?) gets overwritten here as it's only actually set on the EE core request. It'll be read and then saved back to the option in the above.

That all make sense or am I looking at it wrong?

There is an important note on this in that I have to continue let PUE ping ee.com if the installed version is less than the current version, even if `update' has been set with all of the various data we need for the update. This is because the download has a nonce check so if/when someone takes a while to update their download it will fail otherwise, we could rewrite it so that isn't needed but for now I think this gets us most of the way there without rewriting too much of PUE.

Reply:
ya that all sounds fine, I just wanted to make sure that everything is as solid as possible considering this is fairly critical code.

I'm satisfied now that everything you've done is correct and won't result in any foreseeable issues. Great fix 👍

Reply:
ya that all sounds fine, I just wanted to make sure that everything is as solid as possible considering this is fairly critical code.

I'm satisfied now that everything you've done is correct and won't result in any foreseeable issues. Great fix 👍